Add Collector for Active Directory Certificate Services (ADCS)

akrauza opened the pull request.

breed808 commented on collector/adcs.go:
```diff
@@ -0,0 +1,246 @@
+	registerCollector("adcs", ADCSCollectorMethod, "Certification Authority")
+}
+
+type ADCSCollector struct {
```
I don't believe this type should be public

akrauza replied:
Ack. My apologies -- this is one of my first projects in GoLang and I was unfamiliar with the case convention for visibility.

breed808 replied:
It's all good, the case sensitivity rather than a private/public keyword trips a lot of people up :+1:

breed808 commented on collector/adcs.go:
```diff
@@ -0,0 +1,246 @@
+}
+
+// ADCSCollectorMethod ...
+func ADCSCollectorMethod() (Collector, error) {
```
Same with the `ADCSCollector` type: I don't think this needs to be public.

breed808 commented on collector/adcs.go:
```diff
@@ -0,0 +1,246 @@
+func ADCSCollectorMethod() (Collector, error) {
+	const subsystem = "adcs"
+	return &ADCSCollector{
+		RequestsPerSecond: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "requests_total"),
+			"Certificate requests processed per second",
+			[]string{"cert_template"},
+			nil,
```
`requests_total` indicates this is the total number of requests, but the description states this metric is a per-second rate.

There's a number of other metrics with a similar issue, but one comment thread is adequate for this.

akrauza replied:
Metric description updated due to Windows reporting the *last* known value, not an average, etc.

breed808 commented on collector/adcs.go:
```diff
@@ -0,0 +1,246 @@
+			"Certificate requests processed per second",
+			[]string{"cert_template"},
+			nil,
+		),
+		RequestProcessingTime: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "request_processing_time_seconds"),
+			"Time elapsed per request",
+			[]string{"cert_template"},
+			nil,
+		),
```
Does this metric represent the total request processing time for a template, or the average time spent on each request?

akrauza replied:
It is the *last* reported value by Windows for each template. Have updated the metric description to read "Last time elapsed for..."

akrauza replied:
To further clarify, there is a label of `_Total` which would report the *last* known time across *all* certificate templates.
```
# HELP windows_adcs_request_processing_time Time elapsed per request
# TYPE windows_adcs_request_processing_time gauge
windows_adcs_request_processing_time{cert_template="Administrator"} 0
windows_adcs_request_processing_time{cert_template="CEPEncryption"} 0
windows_adcs_request_processing_time{cert_template="DirectoryEmailReplication"} 0
windows_adcs_request_processing_time{cert_template="DomainController"} 11.294
windows_adcs_request_processing_time{cert_template="DomainControllerAuthentication"} 0
windows_adcs_request_processing_time{cert_template="EFS"} 0
windows_adcs_request_processing_time{cert_template="EFSRecovery"} 0
windows_adcs_request_processing_time{cert_template="EnrollmentAgentOffline"} 0
windows_adcs_request_processing_time{cert_template="IPSECIntermediateOffline"} 0
windows_adcs_request_processing_time{cert_template="KerberosAuthentication"} 11.419
windows_adcs_request_processing_time{cert_template="Machine"} 0
windows_adcs_request_processing_time{cert_template="SubCA"} 0
windows_adcs_request_processing_time{cert_template="User"} 0
windows_adcs_request_processing_time{cert_template="WebServer"} 0
windows_adcs_request_processing_time{cert_template="Workstation"} 0
windows_adcs_request_processing_time{cert_template="_Total"} 11.294
```

breed808 replied:
I think this is representing the total time spent processing requests, rather than the total time of the most recent request.

Does this metric increment over time, or increase/decrease with each reading?

akrauza replied:
I confirmed using Performance Monitor that the metric reports the time for the last request, and therefore can increase/decrease as requests are processed by ADCS. I believe the screenshot I attached below shows this.

breed808 replied:
Ah, don't mind me then. I had assumed it was a counter of some kind, which is common with the perflib metrics.

breed808 commented on collector/adcs.go:
```diff
@@ -0,0 +1,246 @@
+	return nil, nil
+}
+
+func milliSecToSecADCS(t float64) float64 {
+	return t / 1000
+}
```
There's a duplicate `milliSecToSec` function in the `remote_fx` exporter.

We could move it to the `collector/collector.go` file and re-use to avoid duplication.

akrauza replied:
Moved to common `collector/collector.go`

breed808 reviewed (changes requested):
Thanks for submitting this Austin! I'm not able to test the collector myself at the moment, but I've left a few queries regarding the metric descriptions and a few other minor issues.

Let me know if any clarification is needed.

akrauza commented:
@breed808 Hey Ben! Thanks for the feedback on the above. Made the changes as requested. Let me know how it looks :)

breed808 commented:
Thanks Austin, I've closed all but one of the comment threads.

Once the metric names/descriptions are in order, you'll then need to update the collector documentation.

At the end we might need to squash or rebase the commits to keep the history tidy, but that's more a nitpick than an issue.

akrauza commented:
Hey @breed808 ! Thank you for your help and review. I updated the documentation and squashed the commit down into one commit message of `Initial commit for ADCS collector` with hash `a89b5377`.

breed808 reviewed (approved):
Looks good!

Thanks for making those amendments Austin :+1:
